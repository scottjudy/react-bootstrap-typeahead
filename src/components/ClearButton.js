// @flow

import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';

import { isSizeLarge, noop } from '../utils';

import { sizeType } from '../propTypes';
import type { EventHandler, KeyboardEventHandler, Size } from '../types';

const propTypes = {
  label: PropTypes.string,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
  size: sizeType,
};

const defaultProps = {
  label: 'Clear',
  onClick: noop,
  onKeyDown: noop,
};

type Props = {
  className?: string,
  label: string,
  onClick: EventHandler<HTMLButtonElement>,
  onKeyDown: KeyboardEventHandler<HTMLButtonElement>,
  size?: Size,
};

/**
 * ClearButton
 *
 * http://getbootstrap.com/css/#helper-classes-close
 */
const ClearButton = ({
  className,
  label,
  onClick,
  onKeyDown,
  size,
  ...props
}: Props) => (
  <button
    {...props}
    aria-label={label}
    className={cx(
      'close',
      'rbt-close',
      {
        'rbt-close-lg': isSizeLarge(size),
      },
      className
    )}
    onClick={(e: SyntheticEvent<HTMLButtonElement>) => {
      e.stopPropagation();
      onClick(e);
    }}
    onKeyDown={(e: SyntheticKeyboardEvent<HTMLButtonElement>) => {
      // Prevent browser from navigating back.
      if (e.key === 'Backspace') {
        e.preventDefault();
      }
      onKeyDown(e);
    }}
    type="button">
    <span aria-hidden="true">&times;</span>
    <span className="sr-only visually-hidden">{label}</span>
  </button>
);

ClearButton.propTypes = propTypes;
ClearButton.defaultProps = defaultProps;

export default ClearButton;
