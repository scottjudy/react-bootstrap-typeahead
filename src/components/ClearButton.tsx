import cx from 'classnames';
import PropTypes from 'prop-types';
import React, {
  HTMLProps,
  KeyboardEvent,
  KeyboardEventHandler,
  MouseEvent,
  MouseEventHandler,
} from 'react';

import type { Size } from '../types';
import { isSizeLarge, noop } from '../utils';
import { sizeType } from '../propTypes';

const propTypes = {
  label: PropTypes.string,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
  size: sizeType,
};

const defaultProps = {
  label: 'Clear',
  onClick: noop,
  onKeyDown: noop,
};

interface ClearButtonProps extends Omit<HTMLProps<HTMLButtonElement>, 'size'> {
  label?: string;
  onClick: MouseEventHandler<HTMLButtonElement>;
  onKeyDown: KeyboardEventHandler<HTMLButtonElement>;
  size?: Size;
}

/**
 * ClearButton
 *
 * http://getbootstrap.com/css/#helper-classes-close
 */
const ClearButton = ({
  className,
  label,
  onClick,
  onKeyDown,
  size,
  ...props
}: ClearButtonProps): JSX.Element => (
  <button
    {...props}
    aria-label={label}
    className={cx(
      'close',
      'rbt-close',
      {
        'rbt-close-lg': isSizeLarge(size),
      },
      className
    )}
    onClick={(e: MouseEvent<HTMLButtonElement>) => {
      e.stopPropagation();
      onClick(e);
    }}
    onKeyDown={(e: KeyboardEvent<HTMLButtonElement>) => {
      // Prevent browser from navigating back.
      if (e.key === 'Backspace') {
        e.preventDefault();
      }
      onKeyDown(e);
    }}
    type="button">
    <span aria-hidden="true">&times;</span>
    <span className="sr-only visually-hidden">{label}</span>
  </button>
);

ClearButton.propTypes = propTypes;
ClearButton.defaultProps = defaultProps;

export default ClearButton;
