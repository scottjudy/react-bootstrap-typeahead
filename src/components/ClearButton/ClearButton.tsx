import cx from 'classnames';
import PropTypes from 'prop-types';
import React, { HTMLProps, KeyboardEvent, MouseEvent } from 'react';

import type { Size } from '../../types';
import { isSizeLarge, isSizeSmall } from '../../utils';

import { sizeType } from '../../propTypes';

const propTypes = {
  label: PropTypes.string,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
  size: sizeType,
};

const defaultProps = {
  label: 'Clear',
};

export interface ClearButtonProps
  extends Omit<HTMLProps<HTMLButtonElement>, 'size'> {
  label: string;
  size?: Size;
}

/**
 * ClearButton
 *
 * http://getbootstrap.com/css/#helper-classes-close
 */
const ClearButton = ({
  className,
  label,
  onClick,
  onKeyDown,
  size,
  ...props
}: ClearButtonProps): JSX.Element => (
  <button
    {...props}
    aria-label={label}
    className={cx(
      'close',
      'btn-close',
      'rbt-close',
      {
        'rbt-close-lg': isSizeLarge(size),
        'rbt-close-sm': isSizeSmall(size),
      },
      className
    )}
    onClick={(e: MouseEvent<HTMLButtonElement>) => {
      e.stopPropagation();
      onClick && onClick(e);
    }}
    onKeyDown={(e: KeyboardEvent<HTMLButtonElement>) => {
      // Prevent browser from navigating back.
      if (e.key === 'Backspace') {
        e.preventDefault();
      }
      onKeyDown && onKeyDown(e);
    }}
    type="button">
    <span aria-hidden="true" className="rbt-close-content">
      &times;
    </span>
    <span className="sr-only visually-hidden">{label}</span>
  </button>
);

ClearButton.propTypes = propTypes;
ClearButton.defaultProps = defaultProps;

export default ClearButton;
