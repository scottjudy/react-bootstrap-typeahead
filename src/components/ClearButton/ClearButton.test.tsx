import React from 'react';

import * as stories from './ClearButton.stories';

import {
  composeStories,
  fireEvent,
  generateSnapshots,
  render,
  screen,
  userEvent,
} from '../../tests/helpers';

const { Default, Large } = composeStories(stories);

describe('<ClearButton>', () => {
  generateSnapshots(stories);

  it('renders a default clear button', () => {
    render(<Default />);
    expect(screen.getByRole('button').className).toBe('close rbt-close');
  });

  it('renders a large clear button', () => {
    render(<Large />);
    expect(screen.getByRole('button').className).toContain('rbt-close-lg');
  });

  it('registers a click', () => {
    const onClick = jest.fn();
    render(<Default onClick={onClick} />);

    const button = screen.getByRole('button');
    userEvent.click(button);

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('prevents the default backspace behavior', () => {
    const onKeyDown = jest.fn();
    let isDefault;

    render(<Default onKeyDown={onKeyDown} />);

    const button = screen.getByRole('button');

    isDefault = fireEvent.keyDown(button, {
      key: 'Backspace',
    });

    expect(onKeyDown).toHaveBeenCalledTimes(1);
    expect(isDefault).toBe(false);

    isDefault = fireEvent.keyDown(button, {
      key: 'Enter',
    });

    expect(onKeyDown).toHaveBeenCalledTimes(2);
    expect(isDefault).toBe(true);
  });
});
