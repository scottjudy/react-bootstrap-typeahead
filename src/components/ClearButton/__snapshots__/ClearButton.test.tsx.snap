// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClearButton> Default story renders snapshot 1`] = `
<button
  aria-label="Clear"
  className="close btn-close rbt-close"
  onClick={[Function]}
  onKeyDown={[Function]}
  type="button"
>
  <span
    aria-hidden="true"
    className="rbt-close-content"
  >
    ×
  </span>
  <span
    className="sr-only visually-hidden"
  >
    Clear
  </span>
</button>
`;

exports[`<ClearButton> Large story renders snapshot 1`] = `
<button
  aria-label="Clear"
  className="close btn-close rbt-close rbt-close-lg"
  onClick={[Function]}
  onKeyDown={[Function]}
  type="button"
>
  <span
    aria-hidden="true"
    className="rbt-close-content"
  >
    ×
  </span>
  <span
    className="sr-only visually-hidden"
  >
    Clear
  </span>
</button>
`;
